File: 백준/Gold/19238. 스타트 택시/스타트 택시.cc
#include <stdio.h>
#define M 100

int zx[4] = { 0,-1,0,1 }, zy[4] = { -1,0,1,0 };
int arr[M][M], pass[M*5][5];
int sx, sy;

int bfs(int n, int x, int y, int tx, int ty) {
    int st, ed;
    int count, zzx, zzy;
    int chk[M][M] = { 0, };
    int q[M*M*M][3];
    int i;

    st = ed = -1;
    q[++st][0] = x; q[st][1] = y; q[st][2] = 0;

    while (st != ed) {
        x = q[++ed][0]; y = q[ed][1]; count = q[ed][2];
        chk[x][y] = 1;

        if (tx == x && ty == y) return count;

        for (i = 0; i < 4; i++) {
            zzx = x + zx[i]; zzy = y + zy[i];

            if (zzx<1 || zzy <1 || zzx > n || zzy > n ||
                arr[zzx][zzy] || chk[zzx][zzy]) continue;

            q[++st][0] = zzx; q[st][1] = zzy; q[st][2] = count + 1;
            chk[zzx][zzy] = 1;
        }
    }

    return -1;
}

int main()
{
    int n, m, k;
    int i, j;
    int charge;
    int min, point;
    int chk[M*5] = { 0, };

    scanf("%d%d%d", &n, &m, &k);

    for (i = 1; i <= n; i++) {
        for (j = 1; j <= n; j++) {
            scanf("%d", &arr[i][j]);
        }
    }

    scanf("%d%d", &sx, &sy);

    for (i = 0; i < m; i++) {
        scanf("%d%d%d%d", &pass[i][0], &pass[i][1], &pass[i][2], &pass[i][3]);
        pass[i][4] = bfs(n, pass[i][0], pass[i][1], pass[i][2], pass[i][3]);
    }

    for (i = 0; i < m; i++) {
        min = 99999999;
        point = 0;
        for (j = 0; j < m; j++) {
            if (chk[j]) continue;
            charge = bfs(n, sx, sy, pass[j][0], pass[j][1]);

            if (charge == -1) {
                k = -99999999;
                point = 0;
                break;
            }

            if (min >= charge) {
                if (min == charge) {
                    if (point != -1) {
                        if (pass[j][0] < pass[point][0]) {
                            point = j;
                        }
                        if (pass[j][0] == pass[point][0] && pass[j][1] < pass[point][1]) {
                            point = j;
                        }
                    }
                    else {
                        point = j;
                    }
                }
                else {
                    min = charge;
                    point = j;
                }
            }
        }

        chk[point] = 1;
        k = k - min - pass[point][4];

        if (k < 0) {
            k = -1; break;
        }
        else {
            k = k + pass[point][4] * 2;
        }

        sx = pass[point][2]; sy = pass[point][3];
    }

    printf("%d", k);
    return 0;
}

File: 백준/Gold/19238. 스타트 택시/스타트 택시.py
import sys
from collections import deque

my_map = []
my_guest = []
direction = [[1,0], [-1,0], [0,1], [0,-1]]

def find_guest(sx, sy, n) :
    global my_guest
    queue = deque()
    queue.append([sx,sy,0])
    
    chk = [[] for _ in range(n)]

    for _ in range(n) :
        my_chk = []
        my_chk = [False for i in range(n)]
        chk[_] = my_chk

    chk[sx][sy]=True
    min_ = 5000
    min_x = 5000
    min_y = 5000

    while queue :
        x,y,count = queue.popleft()
        
        if count <= min_ and my_guest[x][y] :
            if min_x > x or (min_x == x and min_y > y) :
                min_x = x
                min_y = y
                min_ = count

        elif count > min_ :
            break

        for zx, zy in direction :
            zzx = x+zx
            zzy = y+zy

            if zzx < 0 or zzx >= n or zzy < 0 or zzy >=n or chk[zzx][zzy] or my_map[zzx][zzy]:
                continue

            chk[zzx][zzy] = True
            queue.append([zzx, zzy, count+1])
    
    if min_ < 5000 :
        return my_guest[min_x][min_y], min_
    else :
        return -1, -1

def bfs(sx, sy, fx, fy, n) :
    global my_map
    queue = deque()
    queue.append([sx,sy,0])
    
    chk = [[] for _ in range(n)]
    for _ in range(n) :
        my_chk = []
        my_chk = [False for i in range(n)]
        chk[_] = my_chk
        
    chk[sx][sy] = True

    while queue :
        x,y,count = queue.popleft()
        
        if x==fx and y==fy :
            return count

        for zx, zy in direction :
            zzx = x+zx
            zzy = y+zy

            if zzx < 0 or zzx >= n or zzy < 0 or zzy >=n or chk[zzx][zzy] or my_map[zzx][zzy]:
                continue

            chk[zzx][zzy] = True
            queue.append([zzx, zzy, count+1])

    return -1

def solution() :
    n, m, fuel = map(int, sys.stdin.readline().rstrip().split(' '))
    global my_map, my_guest

    my_map = [[] for _ in range(n)]
    my_guest = [[] for _ in range(n)]

    for _ in range(n):
        my_li = sys.stdin.readline().rstrip().split(" ")
        my_li = [int(i) for i in my_li]
        my_map[_] = my_li

        my_temp = [0 for i in range(n)]
        my_guest[_] = my_temp

    sx, sy = map(int, sys.stdin.readline().rstrip().split(' '))
    sx-=1
    sy-=1

    guest_arr = []
    guest_arr = [[] for _ in range(m)]
    flag = False

    for _ in range(m) :
        guest_temp = []
        guest_temp = [[] for _ in range(5)]
        guest_temp[0], guest_temp[1], guest_temp[2], guest_temp[3] = map(int, sys.stdin.readline().rstrip().split(' '))

        for i in range(4) :
            guest_temp[i]-=1
        
        guest_temp[4] = bfs(guest_temp[0], guest_temp[1], guest_temp[2], guest_temp[3], n)

        if guest_temp[4] == -1 :
            flag = True

        my_guest[guest_temp[0]][guest_temp[1]] = (_+1)
        guest_arr[_] = guest_temp

    if flag :
        fuel = -1
    else :
        for i in range(m) :
            point, charge = find_guest(sx,sy,n)

            if point == -1 and charge == -1 :
                fuel = -1
                break

            point -=1
            my_guest[guest_arr[point][0]][guest_arr[point][1]] = 0
            fuel = fuel - charge - guest_arr[point][4]
            
            if fuel < 0 :
                fuel = -1
                break
            else :
                fuel = fuel + guest_arr[point][4] * 2

            sx = guest_arr[point][2]
            sy = guest_arr[point][3]

    print(fuel)

solution()